From 656c2ec7cffdf20a51d272a429a1c3414d152a5d Mon Sep 17 00:00:00 2001
From: Techcable <Techcable@outlook.com>
Date: Tue, 6 Jan 2015 22:12:31 -0600
Subject: [PATCH] POM changes

Check our github versioning

diff --git a/pom.xml b/pom.xml
index 043b7b2..4cf5bbd 100644
--- a/pom.xml
+++ b/pom.xml
@@ -3,18 +3,18 @@
     <modelVersion>4.0.0</modelVersion>
 
     <parent>
-        <groupId>com.destroystokyo.paper</groupId>
-        <artifactId>paper-parent</artifactId>
+        <groupId>net.techcable.tacospigot</groupId>
+        <artifactId>parent</artifactId>
         <version>dev-SNAPSHOT</version>
     </parent>
 
-    <groupId>com.destroystokyo.paper</groupId>
-    <artifactId>paper-api</artifactId>
+    <groupId>net.techcable.tacospigot</groupId>
+    <artifactId>api</artifactId>
     <version>1.9.2-R0.1-SNAPSHOT</version>
     <packaging>jar</packaging>
 
-    <name>Paper-API</name>
-    <url>https://github.com/PaperSpigot/Paper</url>
+    <name>TacoSpigot-API</name>
+    <url>https://github.com/TacoSpigot/TacoSpigot</url>
     <description>An enhanced plugin API for Minecraft servers.</description>
 
     <properties>
diff --git a/src/main/java/org/bukkit/command/defaults/VersionCommand.java b/src/main/java/org/bukkit/command/defaults/VersionCommand.java
index 6d32e1d..5d1a5ac 100644
--- a/src/main/java/org/bukkit/command/defaults/VersionCommand.java
+++ b/src/main/java/org/bukkit/command/defaults/VersionCommand.java
@@ -27,6 +27,14 @@ import java.util.logging.Logger;
 import org.json.simple.JSONObject;
 import org.json.simple.parser.JSONParser;
 import org.json.simple.parser.ParseException;
+// TacoSpigot start
+import com.google.gson.JsonParser;
+import com.google.gson.stream.JsonReader;
+import java.io.InputStreamReader;
+import java.net.HttpURLConnection;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+// TacoSpigot end
 
 // Paper start
 import java.io.InputStreamReader;
@@ -191,7 +199,25 @@ public class VersionCommand extends BukkitCommand {
     private void obtainVersion() {
         String version = Bukkit.getVersion();
         if (version == null) version = "Custom";
-        if (version.startsWith("git-Paper-")) {
+        // TacoSpigot start
+        if (version.startsWith("git-TacoSpigot-")) {
+            String[] parts = version.substring("git-TacoSpigot-".length()).split("[-\\s]");
+            int distance = getDistance(null, parts[0]);
+            switch (distance) {
+                case -1:
+                    setVersionMessage("Error obtaining version information");
+                    break;
+                case 0:
+                    setVersionMessage("You are running the latest version");
+                    break;
+                case -2:
+                    setVersionMessage("Unknown version");
+                    break;
+                default:
+                    setVersionMessage("You are " + distance + " version(s) behind");
+            }
+        } else if (version.startsWith("git-Paper-")) {
+            // TacoSpigot end
             String[] parts = version.substring("git-Paper-".length()).split("[-\\s]");
             int distance = getDistance(null, parts[0]);
             switch (distance) {
@@ -243,13 +269,17 @@ public class VersionCommand extends BukkitCommand {
 
     // Paper start
     private static int getDistance(String repo, String verInfo) {
+        // TacoSpigot start - remove jenkins check
+        /*
         try {
             int currentVer = Integer.decode(verInfo);
             return getFromJenkins(currentVer);
         } catch (NumberFormatException ex) {
+            */
             verInfo = verInfo.replace("\"", "");
-            return getFromRepo("PaperMC/Paper", verInfo);
-        }
+            return getFromRepo("TacoSpigot/TacoSpigot", verInfo); // 'PaperMC/Paper' -> 'TacoSpigot/TacoSpigot'
+        //}
+        // TacoSpigot end
             /*
             BufferedReader reader = Resources.asCharSource(
                     new URL("https://hub.spigotmc.org/stash/rest/api/1.0/projects/SPIGOT/repos/" + repo + "/commits?since=" + URLEncoder.encode(hash, "UTF-8") + "&withCounts=true"),
@@ -267,6 +297,8 @@ public class VersionCommand extends BukkitCommand {
             */
     }
 
+    // TacoSpigot start - remove jenkins check
+    /*
     private static int getFromJenkins(int currentVer) {
         try {
             BufferedReader reader = Resources.asCharSource(
@@ -287,6 +319,8 @@ public class VersionCommand extends BukkitCommand {
             return -1;
         }
     }
+    */
+    // TacoSpigot end
 
     // Contributed by Techcable <Techcable@outlook.com> in GH PR #65
     private static final String BRANCH = "master";
-- 
2.7.4

