From e1694e1ca7f861d88b01bb4ed26c67fefae3a04f Mon Sep 17 00:00:00 2001
From: Techcable <Techcable@outlook.com>
Date: Fri, 25 Sep 2015 13:48:23 -0700
Subject: [PATCH] long map ids


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 584fe11..3619014 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -518,6 +518,12 @@ public final class Bukkit {
         return server.getMap(id);
     }
 
+    // TacoSpigot start - api for long map ids
+    public static MapView getMap(long highId) {
+        return server.getMap(highId);
+    }
+    // TacoSpigot end
+
     /**
      * Create a new map with an automatically assigned ID.
      *
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 26acdda..581086b 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -427,6 +427,14 @@ public interface Server extends PluginMessageRecipient {
     public MapView getMap(short id);
 
     /**
+     * Gets the map from the given high ID.
+     *
+     * @param highId the id of the map to get
+     * @return a map view if it exists, or null
+     */
+    public MapView getMap(long highId);
+
+    /**
      * Create a new map with an automatically assigned ID.
      *
      * @param world the world the map will belong to
diff --git a/src/main/java/org/bukkit/inventory/meta/MapMeta.java b/src/main/java/org/bukkit/inventory/meta/MapMeta.java
index fb5c297..1b023c7 100644
--- a/src/main/java/org/bukkit/inventory/meta/MapMeta.java
+++ b/src/main/java/org/bukkit/inventory/meta/MapMeta.java
@@ -20,4 +20,11 @@ public interface MapMeta extends ItemMeta {
     void setScaling(boolean value);
 
     MapMeta clone();
+
+    // TacoSpigot start - allow plugins to set the id to use
+    /**
+     * Set the id of the map view to use
+     */
+    public void setId(long id);
+    // TacoSpigot end
 }
diff --git a/src/main/java/org/bukkit/map/MapView.java b/src/main/java/org/bukkit/map/MapView.java
index 65c4159..e4d3b83 100644
--- a/src/main/java/org/bukkit/map/MapView.java
+++ b/src/main/java/org/bukkit/map/MapView.java
@@ -65,6 +65,15 @@ public interface MapView {
     @Deprecated
     public short getId();
 
+    // TacoSpigot start - expose api for long map ids
+    /**
+     * Get the ID of this map item.
+     *
+     * @return the map id
+     */
+    public long getLongId();
+    // TacoSpigot end
+
     /**
      * Check whether this map is virtual. A map is virtual if its lowermost
      * MapRenderer is plugin-provided.
-- 
2.5.2.windows.2

