From 9cec069e806747dd0112e1ee8422ec305d0da305 Mon Sep 17 00:00:00 2001
From: Techcable <Techcable@outlook.com>
Date: Sat, 26 Sep 2015 01:56:58 -0700
Subject: [PATCH] Add AsyncChunkPreLoadEvent


diff --git a/src/main/java/net/techcable/tacospigot/event/world/AsyncChunkPreLoadEvent.java b/src/main/java/net/techcable/tacospigot/event/world/AsyncChunkPreLoadEvent.java
new file mode 100644
index 0000000..ea9accf
--- /dev/null
+++ b/src/main/java/net/techcable/tacospigot/event/world/AsyncChunkPreLoadEvent.java
@@ -0,0 +1,44 @@
+package net.techcable.tacospigot.event.world;
+
+import org.bukkit.Bukkit;
+import org.bukkit.Chunk;
+import org.bukkit.ChunkSnapshot;
+import org.bukkit.event.Event;
+import org.bukkit.event.HandlerList;
+
+/**
+ * Fired before the chunk is loaded
+ * <p/>
+ * May be fired multiple times per chunk, from any thread (including the main thread)
+ */
+public class AsyncChunkPreLoadEvent extends Event {
+    private final ChunkSnapshot data;
+
+    private static final HandlerList handlerList = new HandlerList();
+
+    public AsyncChunkPreLoadEvent(Chunk chunk) {
+        this(chunk.getChunkSnapshot());
+    }
+
+    public AsyncChunkPreLoadEvent(ChunkSnapshot data) {
+        super(!Bukkit.isPrimaryThread());
+        this.data = data;
+    }
+
+    public int getX() {
+        return data.getX();
+    }
+
+    public int getZ() {
+        return data.getZ();
+    }
+
+    public static HandlerList getHandlerList() {
+        return handlerList;
+    }
+
+    @Override
+    public HandlerList getHandlers() {
+        return getHandlerList();
+    }
+}
\ No newline at end of file
-- 
2.5.2.windows.2

