From 3482fedd952627dfb8b7563df3ebdd011151fa61 Mon Sep 17 00:00:00 2001
From: Techcable <Techcable@outlook.com>
Date: Fri, 7 Aug 2015 19:31:31 -0700
Subject: [PATCH] Fix head display issues

Log issues with head lookups.

diff --git a/src/main/java/net/minecraft/server/TileEntitySkull.java b/src/main/java/net/minecraft/server/TileEntitySkull.java
index 58014c5..7f53889 100644
--- a/src/main/java/net/minecraft/server/TileEntitySkull.java
+++ b/src/main/java/net/minecraft/server/TileEntitySkull.java
@@ -18,6 +18,10 @@ import com.google.common.util.concurrent.ThreadFactoryBuilder;
 import com.mojang.authlib.Agent;
 import com.mojang.authlib.ProfileLookupCallback;
 // Spigot end
+// TacSpigot start
+import java.util.logging.Level;
+import java.util.logging.Logger;
+// TacoSpigot end
 
 public class TileEntitySkull extends TileEntity {
 
@@ -30,6 +34,7 @@ public class TileEntitySkull extends TileEntity {
                     .setNameFormat("Head Conversion Thread - %1$d")
                     .build()
     );
+    private static long lastMsg;// TacoSpigot
     public static final LoadingCache<String, GameProfile> skinCache = CacheBuilder.newBuilder()
             .maximumSize( 5000 )
             .expireAfterAccess( 60, TimeUnit.MINUTES )
@@ -48,12 +53,18 @@ public class TileEntitySkull extends TileEntity {
 
                         @Override
                         public void onProfileLookupFailed(GameProfile gp, Exception excptn) {
+                            // TacoSpigot start - log errors
+                            if (excptn != null && System.currentTimeMillis() - lastMsg > TimeUnit.MINUTES.toMillis(5)) {
+                                Logger logger = Logger.getLogger("");
+                                logger.log(Level.WARNING, "Unable to lookup profile", excptn);
+                                lastMsg = System.currentTimeMillis();
+                            }
+                            // TacoSpigot end
                             profiles[0] = gp;
                         }
                     };
 
                     MinecraftServer.getServer().getGameProfileRepository().findProfilesByNames(new String[] { key }, Agent.MINECRAFT, gameProfileLookup);
-
                     GameProfile profile = profiles[ 0 ];
                     if (profile == null) {
                         UUID uuid = EntityHuman.a(new GameProfile(null, key));
-- 
2.4.6.windows.1.812.gd1b00d3

