From cf732be0b193cfafba49ad32b5e85e9068526963 Mon Sep 17 00:00:00 2001
From: Techcable <Techcable@outlook.com>
Date: Fri, 7 Aug 2015 19:31:31 -0700
Subject: [PATCH] Fix head display issues

Use uuid api.

diff --git a/src/main/java/net/minecraft/server/TileEntitySkull.java b/src/main/java/net/minecraft/server/TileEntitySkull.java
index 58014c5..a397047 100644
--- a/src/main/java/net/minecraft/server/TileEntitySkull.java
+++ b/src/main/java/net/minecraft/server/TileEntitySkull.java
@@ -14,6 +14,8 @@ import java.util.concurrent.Executor;
 import java.util.concurrent.Executors;
 import java.util.concurrent.TimeUnit;
 
+import net.techcable.tacospigot.uuid.ProfileHelper;
+
 import com.google.common.util.concurrent.ThreadFactoryBuilder;
 import com.mojang.authlib.Agent;
 import com.mojang.authlib.ProfileLookupCallback;
@@ -38,6 +40,8 @@ public class TileEntitySkull extends TileEntity {
                 @Override
                 public GameProfile load(String key) throws Exception
                 {
+                    // TacoSpigot start - use uuid api
+                    /*
                     final GameProfile[] profiles = new GameProfile[1];
                     ProfileLookupCallback gameProfileLookup = new ProfileLookupCallback() {
 
@@ -53,13 +57,14 @@ public class TileEntitySkull extends TileEntity {
                     };
 
                     MinecraftServer.getServer().getGameProfileRepository().findProfilesByNames(new String[] { key }, Agent.MINECRAFT, gameProfileLookup);
-
-                    GameProfile profile = profiles[ 0 ];
+                    */
+                    GameProfile profile = ProfileHelper.lookupNms(key);
+                    // TacoSpigot end
                     if (profile == null) {
                         UUID uuid = EntityHuman.a(new GameProfile(null, key));
                         profile = new GameProfile(uuid, key);
 
-                        gameProfileLookup.onProfileLookupSucceeded(profile);
+                        // gameProfileLookup.onProfileLookupSucceeded(profile); // TacoSpigot
                     } else
                     {
 
@@ -67,7 +72,7 @@ public class TileEntitySkull extends TileEntity {
 
                         if ( property == null )
                         {
-                            profile = MinecraftServer.getServer().aD().fillProfileProperties( profile, true );
+                            ProfileHelper.fillProperties(profile); // TacoSpigot - use uuid api to fill profiles
                         }
                     }
 
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSkull.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSkull.java
index ce5425f..2749999 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSkull.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSkull.java
@@ -5,6 +5,7 @@ import java.util.Map;
 import net.minecraft.server.GameProfileSerializer;
 import net.minecraft.server.NBTBase;
 import net.minecraft.server.NBTTagCompound;
+import net.techcable.tacospigot.uuid.ProfileHelper; // TacoSpigot
 
 import org.bukkit.Material;
 import org.bukkit.configuration.serialization.DelegateDeserialization;
@@ -132,7 +133,7 @@ class CraftMetaSkull extends CraftMetaItem implements SkullMeta {
         if (name == null) {
             profile = null;
         } else {
-            profile = new GameProfile(null, name);
+            profile = ProfileHelper.lookupNms(name); // TacoSpigot - use uuid api
         }
 
         return true;
-- 
2.4.6.windows.1

