From 83da27e9be1be6f6fcb0400d2b34f7775c7af6fa Mon Sep 17 00:00:00 2001
From: Techcable <Techcable@outlook.com>
Date: Wed, 23 Sep 2015 12:29:13 -0700
Subject: [PATCH] Add ArrowCollideEvent


diff --git a/src/main/java/net/minecraft/server/EntityArrow.java b/src/main/java/net/minecraft/server/EntityArrow.java
index 37cb17c..d6c6012 100644
--- a/src/main/java/net/minecraft/server/EntityArrow.java
+++ b/src/main/java/net/minecraft/server/EntityArrow.java
@@ -10,6 +10,11 @@ import org.bukkit.entity.LivingEntity;
 import org.bukkit.event.entity.EntityCombustByEntityEvent;
 import org.bukkit.event.player.PlayerPickupArrowEvent;
 // CraftBukkit end
+// TacoSpigot start
+import net.techcable.tacospigot.event.entity.ArrowCollideEvent;
+import org.bukkit.entity.Arrow;
+// TacoSpigot end
+
 
 public abstract class EntityArrow extends Entity implements IProjectile {
 
@@ -197,6 +202,16 @@ public abstract class EntityArrow extends Entity implements IProjectile {
             }
             // Paper end
 
+            // TacoSpigot start - fire collide event
+            if (movingobjectposition != null && movingobjectposition.entity != null) {
+                ArrowCollideEvent event = new ArrowCollideEvent((Arrow) this.getBukkitEntity(), movingobjectposition.entity.getBukkitEntity());
+                this.world.getServer().getPluginManager().callEvent(event);
+                if (event.isCancelled()) {
+                    movingobjectposition = null;
+                }
+            }
+            // TacoSpigot end
+
             if (movingobjectposition != null) {
                 this.a(movingobjectposition);
             }
-- 
2.8.3

