From 2571e430ae12c7eb8f6d6044e6aef00232158437 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Wed, 16 Mar 2016 22:03:58 -0400
Subject: [PATCH] Fix incorrect getNoDamageTicks


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 8bc8e2077..2e2d028af 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -1294,11 +1294,16 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
 
     @Override
     public int getNoDamageTicks() {
+        // TacoSpigot start - fix incorrect calculation of getNoDamageTicks
+        /*
         if (getHandle().invulnerableTicks > 0) {
             return Math.max(getHandle().invulnerableTicks, getHandle().noDamageTicks);
         } else {
             return getHandle().noDamageTicks;
         }
+        */
+        return Math.max(getHandle().invulnerableTicks, Math.max(0, getHandle().noDamageTicks - getHandle().maxNoDamageTicks / 2));
+        // TacoSpigot end
     }
 
     @Override
-- 
2.13.0

