From 5172236989bcedc019ee72c7032390f68bb8fefe Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Tue, 19 Apr 2016 20:30:55 -0400
Subject: [PATCH] Use client ticks to make attack speed more accurate


diff --git a/src/main/java/net/minecraft/server/EntityHuman.java b/src/main/java/net/minecraft/server/EntityHuman.java
index 4510ed805..0650fd0ac 100644
--- a/src/main/java/net/minecraft/server/EntityHuman.java
+++ b/src/main/java/net/minecraft/server/EntityHuman.java
@@ -24,6 +24,8 @@ import org.bukkit.event.player.PlayerVelocityEvent;
 import org.bukkit.util.Vector;
 // CraftBukkit end
 
+import static net.techcable.tacospigot.TacoSpigotConfig.betterPvp; // TacoSpigot
+
 public abstract class EntityHuman extends EntityLiving {
 
     private static final DataWatcherObject<Float> a = DataWatcher.a(EntityHuman.class, DataWatcherRegistry.c);
@@ -68,6 +70,9 @@ public abstract class EntityHuman extends EntityLiving {
     private final ItemCooldown ce;
     @Nullable
     public EntityFishingHook hookedFish;
+
+    public int clientTicksSinceLastAttack; // TacoSpigot
+
     // Paper start
     public boolean affectsSpawning = true;
     // Paper end
@@ -2067,12 +2072,26 @@ public abstract class EntityHuman extends EntityLiving {
 
     public float getCooledAttackStrength(float adjustTicks) { return r(adjustTicks); } // Paper - OBFHELPER
     public float r(float f) {
-        return MathHelper.a(((float) this.aH + f) / this.dG(), 0.0F, 1.0F);
+        // TacoSpigot start - use both client and server ticks
+        if (betterPvp) {
+            return MathHelper.a(
+                (float) (Math.max(
+                     this.aH,
+                     this.clientTicksSinceLastAttack
+                ) + f) / this.dG(),
+                0.0F,
+                1.0F
+            );
+        } else {
+            return MathHelper.a(((float) this.aH + f) / this.dG(), 0.0F, 1.0F);
+        }
+        // TacoSpigot end
     }
 
     public void resetCooldown() { dH(); } // Paper - OBFHELPER
     public void dH() {
         this.aH = 0;
+        this.clientTicksSinceLastAttack = 0; // TacoSpigot
     }
 
     public ItemCooldown getCooldownTracker() {
diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index a8a6e236e..f1a0bb2a1 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -872,6 +872,8 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
                     this.syncPosition();
                 }
 
+                this.player.clientTicksSinceLastAttack++; // TacoSpigot
+
                 if (this.teleportPos != null) {
                     if (this.e - this.A > 20) {
                         this.A = this.e;
diff --git a/src/main/java/net/techcable/tacospigot/TacoSpigotConfig.java b/src/main/java/net/techcable/tacospigot/TacoSpigotConfig.java
index b7699d8c4..22c6009ab 100644
--- a/src/main/java/net/techcable/tacospigot/TacoSpigotConfig.java
+++ b/src/main/java/net/techcable/tacospigot/TacoSpigotConfig.java
@@ -107,4 +107,9 @@ public class TacoSpigotConfig {
     private static void useArraysForBlockStates() {
         useArraysForBlockStates = getBoolean("useArraysForBlockStates", false);
     }
+
+    public static boolean betterPvp;
+    private static void betterPvp() {
+        betterPvp = getBoolean("betterPvp", false);
+    }
 }
-- 
2.15.1.windows.2

