From 117b4b52d7a32e397ab527af4c3a808a99dfed41 Mon Sep 17 00:00:00 2001
From: frash23 <jacob@bytesizedpacket.com>
Date: Thu, 24 Mar 2016 04:07:04 +0100
Subject: [PATCH] Add option to stop redstone firing BlockPhysicsEvent


diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index b7584ca..1f7f546 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -576,7 +576,9 @@ public abstract class World implements IBlockAccess {
             try {
                 // CraftBukkit start
                 CraftWorld world = ((WorldServer) this).getWorld();
-                if (world != null) {
+				// TacoSpigot start - Add config to disable redstone firing BlockPhysicsEvent 
+				if (world != null && (this.tacoSpigotConfig.isRedstoneFireBPE || !(block instanceof BlockRedstoneWire || block instanceof BlockRedstoneTorch || block instanceof BlockRepeater))) {
+				// TacoSpigot end
                     BlockPhysicsEvent event = new BlockPhysicsEvent(world.getBlockAt(blockposition.getX(), blockposition.getY(), blockposition.getZ()), CraftMagicNumbers.getId(block));
                     this.getServer().getPluginManager().callEvent(event);
 
diff --git a/src/main/java/net/techcable/tacospigot/TacoSpigotWorldConfig.java b/src/main/java/net/techcable/tacospigot/TacoSpigotWorldConfig.java
index 3f18f79..0a20be0 100644
--- a/src/main/java/net/techcable/tacospigot/TacoSpigotWorldConfig.java
+++ b/src/main/java/net/techcable/tacospigot/TacoSpigotWorldConfig.java
@@ -63,6 +63,11 @@ public class TacoSpigotWorldConfig {
         return config.getString("world-settings." + worldName + "." + path, config.getString("world-settings.default." + path));
     }
 
+    public boolean isRedstoneFireBPE;
+    private void isRedstoneFireBPE() {
+        isRedstoneFireBPE = getBoolean("redstone-fire-BlockPhysicsEvent", true);
+    }
+
     public boolean isHopperPushBased;
     private void isHopperPushBased() {
         isHopperPushBased = getBoolean("hopper.push-based", true);
-- 
2.7.4

