From 0d727c41620681c443516a513702cd8fa2bc4d6d Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Tue, 19 Apr 2016 20:30:55 -0400
Subject: [PATCH] Use client ticks to make attack speed more accurate


diff --git a/src/main/java/net/minecraft/server/EntityHuman.java b/src/main/java/net/minecraft/server/EntityHuman.java
index 00fe92fa..fda9dffc 100644
--- a/src/main/java/net/minecraft/server/EntityHuman.java
+++ b/src/main/java/net/minecraft/server/EntityHuman.java
@@ -22,6 +22,8 @@ import org.bukkit.event.player.PlayerVelocityEvent;
 import org.bukkit.util.Vector;
 // CraftBukkit end
 
+import static net.techcable.tacospigot.TacoSpigotConfig.betterPvp; // TacoSpigot
+
 public abstract class EntityHuman extends EntityLiving {
 
     private static final DataWatcherObject<Float> a = DataWatcher.a(EntityHuman.class, DataWatcherRegistry.c);
@@ -64,6 +66,9 @@ public abstract class EntityHuman extends EntityLiving {
     private final ItemCooldown bW;
     @Nullable
     public EntityFishingHook hookedFish;
+
+    public int clientTicksSinceLastAttack; // TacoSpigot
+
     public boolean affectsSpawning = true;
 
     // CraftBukkit start
@@ -1922,11 +1927,25 @@ public abstract class EntityHuman extends EntityLiving {
     }
 
     public float n(float f) {
-        return MathHelper.a(((float) this.aE + f) / this.dr(), 0.0F, 1.0F);
+        // TacoSpigot start - use both client and server ticks
+        if (betterPvp) {
+            return MathHelper.a(
+                (float) (Math.max(
+                     this.aE,
+                     this.clientTicksSinceLastAttack
+                ) + f) / this.dr(),
+                0.0F,
+                1.0F
+            );
+        } else {
+            return MathHelper.a(((float) this.aE + f) / this.dr(), 0.0F, 1.0F);
+        }
+        // TacoSpigot end
     }
 
     public void ds() {
         this.aE = 0;
+        this.clientTicksSinceLastAttack = 0; // TacoSpigot
     }
 
     public ItemCooldown getCooldownTracker() {
diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 37a59de5..52928c4a 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -502,6 +502,8 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
                     this.syncPosition();
                 }
 
+                this.player.clientTicksSinceLastAttack++; // TacoSpigot
+
                 if (this.teleportPos != null) {
                     if (this.e - this.A > 20) {
                         this.A = this.e;
diff --git a/src/main/java/net/techcable/tacospigot/TacoSpigotConfig.java b/src/main/java/net/techcable/tacospigot/TacoSpigotConfig.java
index b7699d8c..22c6009a 100644
--- a/src/main/java/net/techcable/tacospigot/TacoSpigotConfig.java
+++ b/src/main/java/net/techcable/tacospigot/TacoSpigotConfig.java
@@ -107,4 +107,9 @@ public class TacoSpigotConfig {
     private static void useArraysForBlockStates() {
         useArraysForBlockStates = getBoolean("useArraysForBlockStates", false);
     }
+
+    public static boolean betterPvp;
+    private static void betterPvp() {
+        betterPvp = getBoolean("betterPvp", false);
+    }
 }
-- 
2.14.2

