From 5cf882017c7a13c6d80e6ff2df1a9465c8c2c99f Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Tue, 19 Apr 2016 20:30:55 -0400
Subject: [PATCH] Use client ticks to make attack speed more accurate


diff --git a/src/main/java/net/minecraft/server/EntityHuman.java b/src/main/java/net/minecraft/server/EntityHuman.java
index 29c2fd4c9..0239435d8 100644
--- a/src/main/java/net/minecraft/server/EntityHuman.java
+++ b/src/main/java/net/minecraft/server/EntityHuman.java
@@ -22,6 +22,8 @@ import org.bukkit.event.player.PlayerVelocityEvent;
 import org.bukkit.util.Vector;
 // CraftBukkit end
 
+import static net.techcable.tacospigot.TacoSpigotConfig.betterPvp; // TacoSpigot
+
 public abstract class EntityHuman extends EntityLiving {
 
     private static final DataWatcherObject<Float> a = DataWatcher.a(EntityHuman.class, DataWatcherRegistry.c);
@@ -65,6 +67,9 @@ public abstract class EntityHuman extends EntityLiving {
     private final ItemCooldown bX;
     @Nullable
     public EntityFishingHook hookedFish;
+
+    public int clientTicksSinceLastAttack; // TacoSpigot
+
     public boolean affectsSpawning = true;
 
     // CraftBukkit start
@@ -1880,16 +1885,32 @@ public abstract class EntityHuman extends EntityLiving {
         this.datawatcher.set(EntityHuman.bu, nbttagcompound);
     }
 
+    public float getAttackSpeed() { return this.dp(); } // TacoSpigot - OBFHELPER
     public float dp() {
         return (float) (1.0D / this.getAttributeInstance(GenericAttributes.g).getValue() * 20.0D);
     }
 
     public float p(float f) {
-        return MathHelper.a(((float) this.aE + f) / this.dp(), 0.0F, 1.0F);
+        // TacoSpigot start - use both client and server ticks
+        if (betterPvp) {
+            return MathHelper.a(
+                (float) (Math.max(
+                     this.getTicksSinceLastSwing(),
+                     this.clientTicksSinceLastAttack
+                ) + f) / this.getAttackSpeed(),
+                0.0F,
+                1.0F
+            );
+        } else {
+            return MathHelper.a(((float) this.getTicksSinceLastSwing() + f) / this.getAttackSpeed(), 0.0F, 1.0F);
+        }
+        // TacoSpigot end
     }
 
+    public int getTicksSinceLastSwing() { return this.aE; } // TacoSpigot - OBFHELPER
     public void dq() {
         this.aE = 0;
+        this.clientTicksSinceLastAttack = 0; // TacoSpigot
     }
 
     public ItemCooldown getCooldownTracker() {
diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 0b6fd67e7..9f431961d 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -474,6 +474,8 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
                     this.syncPosition();
                 }
 
+                this.player.clientTicksSinceLastAttack++; // TacoSpigot
+
                 if (this.teleportPos != null) {
                     if (this.e - this.A > 20) {
                         this.A = this.e;
diff --git a/src/main/java/net/techcable/tacospigot/TacoSpigotConfig.java b/src/main/java/net/techcable/tacospigot/TacoSpigotConfig.java
index b7699d8c4..22c6009ab 100644
--- a/src/main/java/net/techcable/tacospigot/TacoSpigotConfig.java
+++ b/src/main/java/net/techcable/tacospigot/TacoSpigotConfig.java
@@ -107,4 +107,9 @@ public class TacoSpigotConfig {
     private static void useArraysForBlockStates() {
         useArraysForBlockStates = getBoolean("useArraysForBlockStates", false);
     }
+
+    public static boolean betterPvp;
+    private static void betterPvp() {
+        betterPvp = getBoolean("betterPvp", false);
+    }
 }
-- 
2.13.0

